lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ApacheParquetValueDecoders.java

```diff
@@ -371,6 +378,60 @@ public void skip(int n)
         }
     }
 
+    public static final class BinaryApacheParquetValueDecoder
+            implements ValueDecoder<BinaryBuffer>
+    {
+        private final ValuesReader delegate;
+        private final Type trinoType;
+
+        public BinaryApacheParquetValueDecoder(ValuesReader delegate, Type trinoType)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+            this.trinoType = requireNonNull(trinoType, "trinoType is null");
+        }
+
+        @Override
+        public void init(SimpleSliceInputStream input)
+        {
+            initialize(input, delegate);
+        }
+
+        @Override
+        public void read(BinaryBuffer values, int offsetsIndex, int length)
+        {
+            // A dedicated loop for every case bring significant performance benefit
+            if (trinoType instanceof VarcharType && !((VarcharType) trinoType).isUnbounded()) {
```

Comment:
Ideally, the decoder shouldn't be checking for Trino types. Create three separate decoders, instead, and update the factory to return the appropriate one.

Reply:
Refactored into separate decoders